### PR TITLE
18 notifications timeout

### LINKED ISSUE #18: Notifications timeout
**As a** site user 
**I want** a notification **to** disappear after a configurable <timeout>
**In order to** save manual efforts needed to close many notifications.

### PR DESCRIPTION
**Pull request by VadymVadymov**
Closes #18

### CODE REVIEW

File: frontend/src/Component/Error.purs [hunk 1]
```diff
@@ -2,36 +2,60 @@ module Component.Error where
...
+      , HH.span
+          [ HP.classNames [ "font-medium", "text-center" ]
+          ]
+          [ HH.text
+              """Seems like something went wrong! 
+              You may contact our developers or try again later."""
+          ]
+      ]
+
+  ]
```
Comment (Unisay) [inline frontend/src/Component/Error.purs:61]: please configure your vscode to always add a trailing newline

File: frontend/src/Main.purs [hunk 2]
```diff
@@ -23,7 +25,12 @@ main = runHalogenAff do
...
-    when (old /= Just new) do
-      launchAff_ $ router.query $ H.mkTell $ Router.Navigate new
+    runRouter = do
+      router ← runUI ui unit body
+      void $ liftEffect $ matchesWith (RD.parse Route.codec) \old new →
+        when (old /= Just new) do
+          launchAff_ $ router.query $ H.mkTell $ Router.Navigate new
+  runRouter `catchError` \e → do
+    Console.logShow e
+    void $ runUI Error.component unit body
```
Comment (Unisay) [inline frontend/src/Main.purs]: This is not working, please revert.

File: frontend/src/Main.purs [hunk 3]
```diff
@@ -23,7 +25,12 @@ main = runHalogenAff do
   let
     config = { notifications, backendApiUrl: "http://localhost:8081" }
     ui = H.hoist (App.run config) Router.component
-  router ← runUI ui unit body
-  void $ liftEffect $ matchesWith (RD.parse Route.codec) \old new →
-    when (old /= Just new) do
-      launchAff_ $ router.query $ H.mkTell $ Router.Navigate new
```
Comment (Unisay) [inline frontend/src/Main.purs]: The old way is correct

File: frontend/src/Data/Route.purs [hunk 4]
```diff
@@ -27,6 +27,7 @@ data Route
   | Profile Username -- /profile/:username
   | Debug -- /debug 
   | ChatWindow -- /chat
+  | Error -- /oops
```
Comment (Unisay) [inline frontend/src/Data/Route.purs]: `/error` is better than `/oops`

File: frontend/spago.dhall [hunk 5]
```diff
@@ -23,6 +23,7 @@
   , "maybe"
   , "newtype"
   , "ordered-collections"
+  , "partial"
```
Comment (Unisay) [inline frontend/spago.dhall]: Looks like this one isn't used

File: frontend/src/Backend.purs [hunk 6]
```diff
@@ -5,58 +5,62 @@ import Prelude
...
   ⇒ MonadAsk { backendApiUrl ∷ String | r } m
+  ⇒ MonadThrow Error m
   ⇒ Username
   → Password
   → m SignInResponse
 createSession = createSession' AX.request
 
 createSession'
-  ∷ ∀ r m
+  ∷ ∀ r m -- m = EitherT Backend.Error Aff
```
Comment (Unisay) [inline frontend/src/Backend.purs]: Please remove  the comment

File: frontend/src/Component/Router.purs [hunk 7]
```diff
@@ -34,10 +35,12 @@ data Action = Initialize
 type ChildSlots =
   ( notifications ∷ H.OpaqueSlot Unit
   , home ∷ H.OpaqueSlot Unit
+  , error ∷ H.OpaqueSlot Unit
   , signin ∷ H.OpaqueSlot Unit
   , signup ∷ H.OpaqueSlot Unit
   , debug ∷ H.OpaqueSlot Unit
   , chatWindow ∷ H.OpaqueSlot Unit
+  , error ∷ H.OpaqueSlot Unit
```
Comment (Unisay) [inline frontend/src/Component/Router.purs:42]: This one is a duplicate

File: frontend/src/Component/Signin.purs [hunk 8]
```diff
@@ -286,13 +271,18 @@ handleAction = case _ of
...
+          case signInResponse of
+            SignedIn token → do
+              H.modify_ _ { response = Just (SignedIn token) }
+              goTo Route.ChatWindow
+            Forbidden →
+              H.modify_ _ { response = Just Forbidden }
+
+-- Failure str → do
+--   listener ← lift $ asks _.notifications.listener
+--   liftEffect $ HS.notify listener (critical str)
```
Comment (Unisay) [inline frontend/src/Component/Signin.purs]: Please remove commented out code

File: frontend/src/Data/Route.purs [hunk 9]
```diff
@@ -48,11 +50,13 @@ codec = RouteDuplex i o
...
   RouteDuplex i o = root $ G.sum
     { "Home": G.noArgs
+
     , "SignIn": path "signin" G.noArgs
     , "SignUp": path "signup" G.noArgs
     , "Profile": path "profile" (username segment)
     , "Debug": path "debug" G.noArgs
-    , "ChatWindow": path "chatWindow" G.noArgs
+    , "ChatWindow": path "chat" G.noArgs
+    , "Error": path "oops" G.noArgs
```
Comment (Unisay) [inline frontend/src/Data/Route.purs]:
```suggestion
    , "Error": path "error" G.noArgs
```

File: frontend/test/BackendSpec.purs [hunk 10]
```diff
@@ -65,28 +69,27 @@ spec = describe "Backend" do
...
-        SignedUp → fail "Unexpected error was expected"
-        AlreadyRegistered → fail "Unexpected error was expected"
-        Unexpected err →
-          fail ("ServerErrors expected, but got: " <> show err)
-        ServerErrors _err → pure unit
-
+      let
+        server _request = respond badRequest400
+        getErr = createAccountWithConfig server username password email
+      response ← getErr `catchError` checkErrorSignUp "Bad request error"
```
Comment (Unisay) [inline frontend/test/BackendSpec.purs]:
```suggestion
      response ← 
        createAccountWithConfig server username password email
          `catchError` checkErrorSignUp "Bad request error"
```

File: frontend/test/BackendSpec.purs [hunk 11]
```diff
@@ -65,28 +69,27 @@ spec = describe "Backend" do
...
     it "handles request error" do
-      let server _request = pure $ Left AX.RequestFailedError
-      response ← createAccountWithConfig server username password email
-      response `isUnexpectedError` "request failed"
+      let
+        server _request = pure $ Left AX.RequestFailedError
+        getErr = createAccountWithConfig server username password email
+      response ← getErr `catchError` checkErrorSignUp "request failed"
+      isSignedUp response
+    -- response `isUnexpectedError` "request failed"
```
Comment (Unisay) [inline frontend/test/BackendSpec.purs]:
```suggestion
```

File: frontend/test/BackendSpec.purs [hunk 12]
```diff
@@ -65,28 +69,27 @@ spec = describe "Backend" do
...
     it "handles timeout error" do
-      let server _request = pure $ Left AX.TimeoutError
-      response ← createAccountWithConfig server username password email
-      response `isUnexpectedError` "timeout"
+      let
+        server _request = pure $ Left AX.TimeoutError
+        getErr = createAccountWithConfig server username password email
+      response ← getErr `catchError` checkErrorSignUp "timeout"
+      isSignedUp response
+  -- response `isUnexpectedError` "timeout"
```
Comment (Unisay) [inline frontend/test/BackendSpec.purs]:
```suggestion
```

File: frontend/test/BackendSpec.purs [hunk 13]
```diff
@@ -107,67 +110,59 @@ spec = describe "Backend" do
...
     it "handles request error" do
-      let server _request = pure $ Left AX.RequestFailedError
-      response ← createSessionWithConfig server username password
-      response `isFailure` "request failed"
+      let
+        server _request = pure $ Left AX.RequestFailedError
+        getErr = createSessionWithConfig server username password
+      response ← getErr `catchError` checkErrorSignIn "request failed"
+      isSignedIn response
+    -- response `isFailure` "request failed"
```
Comment (Unisay) [inline frontend/test/BackendSpec.purs]:
```suggestion
```

File: frontend/test/BackendSpec.purs [hunk 14]
```diff
@@ -107,67 +110,59 @@ spec = describe "Backend" do
...
     it "handles timeout error" do
-      let server _request = pure $ Left AX.TimeoutError
-      response ← createSessionWithConfig server username password
-      response `isFailure` "timeout"
+      let
+        server _request = pure $ Left AX.TimeoutError
+        getErr = createSessionWithConfig server username password
+      response ← getErr `catchError` checkErrorSignIn "timeout"
+      isSignedIn response
+    -- response `isFailure` "timeout"
```
Comment (Unisay) [inline frontend/test/BackendSpec.purs]:
```suggestion
```

File: frontend/test/BackendSpec.purs [hunk 15]
```diff
@@ -107,67 +110,59 @@ spec = describe "Backend" do
...
   ∷ Transport → Username → Password → Aff SignInResponse
-createSessionWithConfig server username password =
-  App.run { backendApiUrl: "http://localhost/mock" }
-    $ createSession' server username password
+createSessionWithConfig server username password = do
+  notifications ← liftEffect H.create
+  let config = { notifications, backendApiUrl: "http://localhost/mock" }
+  App.run config
+    ( createSession' server username password
+        # AppM.hoistAppM App.BackendError
```
Comment (Unisay) [inline frontend/test/BackendSpec.purs]: I am not sure that you need to hoist backend error, maybe its possible to handle it directly?

Comment (VadymVadymov) [reply frontend/test/BackendSpec.purs]: there was a monadThrow instance conflict between two errors so i believe we need to hoist it there

**Review by Unisay — APPROVED** [review]

File: frontend/test/BackendSpec.purs (hunk 15 above)
Comment (Unisay) [reply frontend/test/BackendSpec.purs]: I removed App and used `ReaderT Config (ExceptT Backend.Error Aff SignInResponse)` instead

**Pull request merged**